### ISSUE TITLE
Waveform in editor has visual delay

### ISSUE THREAD

**Issue by Dopaminos** [ppy/osu#21947]
### Type

Cosmetic

### Bug description

Waveform is visually delayed by ~+15-20ms. Screenshot with some ranked maps waveforms attached. Obviously it shouldn't work like this because it's wrong

### Screenshots or videos

![image](https://user-images.githubusercontent.com/42616718/210085876-365da5e8-f46c-48c8-96d2-7fce75256c37.png)
![image](https://user-images.githubusercontent.com/42616718/210086109-5e170ea5-c664-4215-bb74-12886e0cd1a8.png)

### Version

20221228

### Logs

no need in logs

**Comment by bdach** [comment]
This has been claimed before several times and turned out to be incorrect (https://github.com/ppy/osu/issues/11400#issuecomment-1235618951).

Please link the maps you're testing with for verification.

**Comment by Dopaminos** [comment]
https://osu.ppy.sh/beatmapsets/1839226/discussion/-/generalAll#/3474615
![image](https://user-images.githubusercontent.com/42616718/210086861-eb9a4c1b-215b-4c8b-9070-a9fe0583649b.png)

**Comment by bdach** [comment]
Yeah I dunno I'm not seeing it.

https://osu.ppy.sh/beatmapsets/594170#osu/1256809 @ 03:35:230-03:35:309:

| lazer | audacity |
| :-: | :-: |
| ![1672415487](https://user-images.githubusercontent.com/20418176/210088783-f76b08c9-3701-42df-a2a5-e4f7c2484385.png) | ![1672415537](https://user-images.githubusercontent.com/20418176/210088797-d7898324-56dd-4f78-97d4-e9e1b192ee22.png) |

https://osu.ppy.sh/beatmapsets/1736332#osu/3548649 @ 01:13:419-01:13:488:

| lazer | audacity |
| :-: | :-: |
| ![1672415662](https://user-images.githubusercontent.com/20418176/210089019-50a3d27d-c92b-4436-a2ca-aa6fe443be61.png) | ![1672415739](https://user-images.githubusercontent.com/20418176/210089028-db64be98-2b79-4788-aeff-d2f42da3c65b.png) |

So either both lazer and audacity are wrong, or this information "from experience" is not correct. @Hiviexd care to explain further?

**Comment by Dopaminos** [comment]
Both are wrong here, i'm 100% sure.

In my experience, and some other people who work in Audacity with audio to trim or re-encode it (hi defreeyay and ikikaera), have noticed that Audacity can change the offset of a song for Stable, usually by about -20ms, so I agree with the statement that Lazer and Audacity are wrong.

**Comment by Hiviexd** [comment]
whenever i try to observe the waveform it feels like ticks are earlier than where they're supposed to be visually, mayb it's normal and i'm misunderstanding something? it feels like they should be around the circled area below, because that's usually the start/peak of the sound that we try to time for

![image](https://user-images.githubusercontent.com/62819481/210092108-e1d18233-2541-463b-9ee1-0d71a12e5faf.png)

map: https://osu.ppy.sh/beatmapsets/1860311#taiko/3824662
white tick on 02:33:523

**Comment by IceDynamix** [comment]
the waveform is correct, the issue comes from somewhere else, requires deeper discussion in order to be solved properly and also needs to be addressed before the lazer editor can be fully used as the primary editor

all maps that were timed using stable were timed with audio only, which specifically includes hitsound delay and other sources. other rhythm games whose editors include waveforms show that the waveform is equally off, by about 20-25ms for each map.

to solve this issue, either all maps timed in stable need to receive an online offset of 20-25ms, or the lazer waveform needs to be visually shifted to line up with past timing.

tl;dr all maps are mistimed by 20-25ms in relation to the true waveform

**Comment by peppy** [comment]
See https://github.com/ppy/osu/blob/7bc8908ca9c026fed1d831eb6e58df7624a8d614/osu.Game/Beatmaps/FramedBeatmapClock.cs#L85

This either needs to be applied visually, or moved to another level of implementation (ie. at import / export time) if we want to fix the underlying issue. Probably the first for now, for simplicity.

We've already had discussions regarding this internally, and it was recently fixed for the editor from an audible perspective. The visual display may still need updating if it's not being correctly applied there for whatever reason. Or maybe it doesn't as it should match the actual track. Either way, as mentioned it requires further investigation.

Claiming lazer is "unusable for timing" is a tad misleading though. It should match to the millisecond in audio and gameplay. Please don't use the waveform alone for timing - use your ear.

**Comment by peppy** [comment]
That's what I was also seeing on macOS. I believe previously when I was testing this (and applied it only to windows), I may have been using incorrect assumptions that it was being applied in areas of the game that it wasn't at that point.

I still haven't concluded my investigations here, and in addition, if it's decided that this should be applied across all platforms it may be that we want to move it to a different layer.

**Comment by Trung0246** [comment]
I'm curious if a song that requires 0 offset would sufficient to calibrate?

I think I found some beatmap that seems to have "wrong offset" in lazer but when revert offset back to 0 things looks fit (probably mistimed in stable since `all maps that were timed using stable were timed with audio only, which specifically includes hitsound delay and other sources`). All that in mind here's the beatmap:

https://osu.ppy.sh/beatmapsets/217935 (mistimed by -24 offset, you can see clearly in lazer editor the waveform fits perfectly when offset changed to 0)
https://osu.ppy.sh/beatmapsets/1317808 (mistimed by -18 offset, same as above when changed to 0 offset)
https://osu.ppy.sh/beatmapsets/1846720 (mistimed by -16, same here)

**Comment by peppy** [comment]
Most MP3 encoders themselves add silence to the start. It's not related to this conversation / issue.

Please take discussion elsewhere for now, I don't think anything further is required to be discussed here (next step would be to test each platform further and implement a fix in code)

**Comment by peppy** [comment]
TODO: separately check https://osu.ppy.sh/beatmapsets/1676482#osu/3902488 as noted in #23897, which looks to have 200ms discrepancy.

**Comment by Trung0246** [comment]
Further notes, I think the time rate are misaligned, not just offset delay.

**Comment by i77xd** [comment]
https://github.com/ppy/osu/issues/21947#issuecomment-1367998379
> the waveform is correct, the issue comes from somewhere else, requires deeper discussion in order to be solved properly and also needs to be addressed before the lazer editor can be fully used as the primary editor
> 
> all maps that were timed using stable were timed with audio only, which specifically includes hitsound delay and other sources. other rhythm games whose editors include waveforms show that the waveform is equally off, by about 20-25ms for each map.
> 
> to solve this issue, either all maps timed in stable need to receive an online offset of 20-25ms, or the lazer waveform needs to be visually shifted to line up with past timing.
> 
> tl;dr all maps are mistimed by 20-25ms in relation to the true waveform


I see no one mentioning what I think is the main reason that contributes to mistiming of maps.
Mappers are mistiming maps because of the distorted slowed playback speed (specifically 25%) which they use for timing and finding offset.
Stable doesn't adjust these slower playback speeds by pitch - which results in distortion and the sound appear to start earlier than it actually is.

Here's a demostration using this [map](https://osu.ppy.sh/beatmapsets/203698#mania/481287) which is perfectly timed as shown on waveform in lazer ver. 2023.1224.0 (before the visual delay waveform update)

![image](https://github.com/ppy/osu/assets/62654610/b06fc49e-e02b-4a5d-a913-3067809d6edc)

https://github.com/ppy/osu/assets/62654610/1b225008-78ba-48be-ac13-6fabca190c76

You can clearly hear and see in the video that the slowed speed (default, not adjusted by pitch, distorted) sounds earlier which confuses mappers to set the offset incorrectly.

I managed to adjust the playback speed by pitch by switching audio devices on Windows back and forth. This might be a bug or a hidden feature; regardless, it should have been a visible toggle switch (by pitch being the default) in the editor since the inception of osu! - maps would not have been this much mistimed.

https://github.com/ppy/osu/pull/26136
In a recent lazer update, there was a decision to add a 20ms visual delay to the waveform to compensate for this error mappers do on stable and I am heavily against it and want it to be reverted. The waveform was correct and now it's useless.
